logic/src/commonMain/kotlin/com/wire/kalium/logic/network/SessionManagerImpl.kt

```diff
@@ -31,19 +33,24 @@ class SessionManagerImpl(
     private val serverConfigMapper: ServerConfigMapper = MapperProvider.serverConfigMapper(),
     private val idMapper: IdMapper = MapperProvider.idMapper()
 ) : SessionManager {
-    override fun session(): Pair<SessionDTO, ServerConfigDTO> = sessionRepository.fullAccountInfo(userId).fold({
-        TODO("IMPORTANT! Not yet implemented")
-    }, { account ->
-        val session: SessionDTO = wrapStorageRequest { tokenStorage.getToken(idMapper.toDaoModel(account.info.userId)) }
+
+    private val session: Atomic<SessionDTO?> = Atomic(null)
+    private var serverConfig: Atomic<ServerConfigDTO?> = Atomic(null)
+
+    override fun session(): SessionDTO = session.get() ?: run {
+        wrapStorageRequest { tokenStorage.getToken(idMapper.toDaoModel(userId)) }
             .map { sessionMapper.fromEntityToSessionDTO(it) }
-            .fold({
-                throw IllegalStateException("No token found for user")
-            }, {
-                it
-            })
-        val serverConfig = serverConfigMapper.toDTO(account.serverConfig)
-        session to serverConfig
-    })
+            .onSuccess { session.set(it) }
+        session.get()!!
+    }
+
+    override fun serverConfig(): ServerConfigDTO = serverConfig.get() ?: run {
+        serverConfig.set(sessionRepository.fullAccountInfo(userId)
+            .map { serverConfigMapper.toDTO(it.serverConfig) }
+            .fold({ TODO() }, { it })
```

Comment:
Maybe we want to add a `"Handle set fullAccountInfo on serverConfig"` so if it ever fails we will know where it is failing